text/3579-pki.md

```diff
@@ -0,0 +1,262 @@
+- To prevent the destruction of the quorum, we recommend that no real world event occur in which 5 or more members of the quorum are present; we recommend at least 5 (M-2) members of the quorum remain at geographically diverse locations. (If in the future the quorum threshold is changed, this limit should be changed accordingly to remain M-2).
+- Hardware keys shall be utilized for two factor authentication for Quorum IAM accounts; phone-based or weaker methods of 2FA authentication cannot be used
+
+### Delegate Keys
+
+- Infra Key
+Key for infrastructure team to utilize for server operations. The Infra team may create further delegate keys as necessary; we expect this key to be used for server-side and client-side certificates and other infrastructure-related uses.
+
+- Release Key
+Used to sign Rust nightly and stable releases. rustup can verify that downloaded releases are signed by a key with this role that chains up to the root, and verify that the key and the release signature are recorded via Certificate Transparency / Binary Transparency authenticated logging mechanisms. Another subkey of this may be used for code signing.
+
+- crates.io Key
+Top-level key for the crates.io ecosystem and project. This will be an authoritative key for signing indexes, and potentially signing crate packages directly. (For logistics reasons, crates.io may choose to issue subkeys under this key to perform signing.) Note that even if crates.io signs package files, it still needs to sign indexes for verification as well, because cargo gets some information from the index rather than the crate.
+
+- BORS Key
+Key for bors to sign all git commits it makes. This allows people to use mirrors of our git repositories and verify the authenticity of those mirrors. We will document a procedure to download and update Rust git repositories from a mirror, with verification, while working with the original for issues and pull requests.
+
+### Key Transparency (`pki.rust-lang.org`)
+
+All certificate operations on the Root Key and the top-level Delegate Keys will be logged via Certificate Transparency, on a domain [pki-rust-lang-org][pki-rust-lang-org]. (This prevents issuing or using a Delegate Key surreptitiously.) Each issued key must have proof of the Certificate Transparency record stapled to it.
+
+In addition, we recommend that binary signatures make use of Binary Transparency. (This prevents `crates.io` from serving a signed index or crate file to a subset of requests without logging it publicly.)
+
+We may also wish to set up an OCSP service; we defer this question for evaluation during implementation.
+
+### Rotation & Revocation
+
+- Delegate Keys can be re-issued and re-signed by the Root Key
+- Root Key rotation may require a Rust/Cargo/rustup update to update the local certificate
```

Comment:
Does this mean the Rust toolchain will irrevocably stop working after 7 years since its release?

Reply:
@pietroalbini No. It implies that, for instance, a 7-year-old version of Cargo may not be able to download from crates.io, and a 7-year-old version of rustup may not be able to `rustup update`. People in the position of running a 7-year-old version while still doing such network updates have multiple options: manually update the root keys by some process, or work with folks to get a toolchain point release made, or turn off verification. This is similar to how decade-old OSes with no updates may not be able to browse today's TLS sites without updating their system certificate store somehow.

Reply:
I was mostly referring to being unable to resolve dependencies, yes.

There are crates with fairly long MSRV that need old toolchain versions to ensure they don't break their MSRV. A quick `rg` in my `~/.cargo/registry/src` points out for example the [`paste` crate](https://crates.io/crates/paste), which has an MSRV of 1.31 [and tests that on CI](https://github.com/dtolnay/paste/blob/ed844dc6fe755bcee881bd93cdff5a77038aa49b/.github/workflows/ci.yml#L39-L48). If there is no way to specify new keys, in a year and half it would become way harder to test compatibility.

To be clear this is fairly easy to solve, we just need to add a way to replace the pinned root key (behind proper security warnings).

Reply:
Addressed in the changes we just pushed, with an explanation of the options for people who need to run a toolchain older than 7 years old.